x/sdk/p2p/router.go

```diff
@@ -0,0 +1,224 @@
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	if _, ok := r.handlers[handlerID]; ok {
+		return nil, fmt.Errorf("failed to register handler id %d: %w", handlerID, ErrExistingAppProtocol)
+	}
+
+	client := &Client{
+		handlerID: handlerID,
+		sender:    sender,
+		router:    r,
+	}
+
+	responder := responder{
+		handler: handler,
+		client:  client,
+	}
+
+	r.handlers[handlerID] = responder
+
+	return client, nil
+}
+
+func (r *Router) AppRequest(ctx context.Context, nodeID ids.NodeID, requestID uint32, deadline time.Time, request []byte) error {
+	r.lock.RLock()
+	app, parsedMsg, handler, ok := r.parse(request)
+	r.lock.RUnlock()
+	if !ok {
+		return fmt.Errorf("failed to process app request message for app protocol %d: %w", app, ErrUnregisteredHandler)
+	}
+
+	return handler.AppRequest(ctx, nodeID, requestID, deadline, parsedMsg)
+}
+
+func (r *Router) AppRequestFailed(_ context.Context, nodeID ids.NodeID, requestID uint32) error {
+	r.lock.RLock()
+	pending, ok := r.clearAppRequest(requestID)
+	r.lock.RUnlock()
+	if !ok {
+		return ErrUnrequestedResponse
+	}
+
+	pending.onResponse(nodeID, nil, ErrAppRequestFailed)
+	return nil
+}
+
+func (r *Router) AppResponse(_ context.Context, nodeID ids.NodeID, requestID uint32, response []byte) error {
+	r.lock.RLock()
+	pending, ok := r.clearAppRequest(requestID)
+	r.lock.RUnlock()
+	if !ok {
+		return ErrUnrequestedResponse
+	}
+
+	pending.onResponse(nodeID, response, nil)
+	return nil
+}
+
+func (r *Router) AppGossip(ctx context.Context, nodeID ids.NodeID, gossip []byte) error {
+	r.lock.RLock()
+	app, parsedMsg, handler, ok := r.parse(gossip)
+	r.lock.RUnlock()
+	if !ok {
+		return fmt.Errorf("failed to process gossip message for app protocol %d: %w", app, ErrUnregisteredHandler)
+	}
+
+	return handler.AppGossip(ctx, nodeID, parsedMsg)
+}
+
+func (r *Router) CrossChainAppRequest(
+	ctx context.Context,
+	chainID ids.ID,
+	requestID uint32,
+	deadline time.Time,
+	msg []byte,
+) error {
+	r.lock.RLock()
+	app, parsedMsg, handler, ok := r.parse(msg)
+	r.lock.RUnlock()
+
+	if !ok {
+		return fmt.Errorf("failed to process cross chain app request message for app protocol %d: %w", app, ErrUnregisteredHandler)
+	}
+
+	return handler.CrossChainAppRequest(ctx, chainID, requestID, deadline, parsedMsg)
+}
+
+func (r *Router) CrossChainAppRequestFailed(_ context.Context, chainID ids.ID, requestID uint32) error {
+	r.lock.RLock()
+	pending, ok := r.clearCrossChainAppRequest(requestID)
+	r.lock.RUnlock()
+	if !ok {
+		return ErrUnrequestedResponse
+	}
+
+	pending.onResponse(chainID, nil, ErrAppRequestFailed)
+	return nil
+}
+
+func (r *Router) CrossChainAppResponse(_ context.Context, chainID ids.ID, requestID uint32, response []byte) error {
+	r.lock.RLock()
+	pending, ok := r.clearCrossChainAppRequest(requestID)
+	r.lock.RUnlock()
+	if !ok {
+		return ErrUnrequestedResponse
+	}
+
+	pending.onResponse(chainID, response, nil)
+	return nil
+}
```

Comment:
Returning errors here will be treated as FATAL by the consensus engine - is that intended?

Reply:
I was thinking we would need these errors (unrequested response, invalid request) for migrating any vms that aren't using this sdk currently. During migration, we won't know if a request is/isn't an sdk request, so I was thinking we could handle it with the sdk handler, and if it's an error (unrequested response, non-registered protocol request), we'd forward it to the vm's app handler, and fatal otherwise.